src/main/java/com/musala/atmosphere/client/websocket/ClientDispatcher.java

```diff
@@ -0,0 +1,308 @@
+                    // Nothing to do here.
+                }
+                allocateDeviceRetryCount--;
+            } catch (NoDeviceMatchingTheGivenSelectorException e) {
+                break;
+            }
+        }
+
+        throw new NoAvailableDeviceFoundException();
+    }
+
+    /**
+     * Sends a release device request to the Server and receives a response.
+     *
+     * @param deviceInformation
+     * @throws Exception
+     *         - when an error occurred when trying to release a device
+     */
+    public void releaseDevice(DeviceAllocationInformation deviceInformation) throws Exception {
+        RequestMessage releaseDeviceRequest = new RequestMessage(MessageAction.RELEASE_DEVICE, deviceInformation);
+        releaseDeviceRequest.setSessionId(session.getId());
+
+        ResponseMessage response = sendRequest(releaseDeviceRequest, session);
+        if (response != null && response.getException() != null) {
+            LOGGER.error("An error occurred when trying to release a device.", response.getException());
+            throw response.getException();
+        }
+    }
+
+    /**
+     * Sends a request to get a list of all available devices to the Server and receives a response.
+     *
+     * @return list with serial numbers and models of all available devices
+     * @throws Exception
+     *         - when fails to get a list of the available devices
+     */
+    @SuppressWarnings("unchecked")
+    public List<Pair<String, String>> getAllAvailableDevices() throws Exception {
+        RequestMessage getAllAvailableDevicesRequest = new RequestMessage(MessageAction.GET_ALL_AVAILABLE_DEVICES);
+        getAllAvailableDevicesRequest.setSessionId(session.getId());
+        ResponseMessage response = sendRequest(getAllAvailableDevicesRequest, session);
+        if (response.getMessageAction() == MessageAction.ERROR) {
+            throw response.getException();
+        }
+
+        return (List<Pair<String, String>>) response.getData();
+    }
+
+    /**
+     * Sends a request and waits for a certain time for a response. If the connection is lost it tries to reconnect.
+     */
+    private ResponseMessage sendRequest(RequestMessage request, Session session) {
+        String sessionId = request.getSessionId();
+        Object lockObject = communicationManager.getSynchronizationObject(sessionId);
+
+        String requestJSON = jsonUtil.serialize(request);
+
+        try {
+            session.getBasicRemote().sendText(requestJSON);
+        } catch (IOException e1) {
+            try {
+                connectToServer(serverConnectionProperties.getIp(),
+                                serverConnectionProperties.getPort(),
+                                HANDLE_LOST_CONNECTION_RETRIES);
+            } catch (ServerConnectionFailedException e2) {
+                LOGGER.fatal("Faild to send " + request.getMessageAction() + " request.", e2);
+                throw new ServerConnectionFailedException();
```

Comment:
why do you "hide" e2? Just rethrow it - then you will preserve the message. Is it really needed to catch and throw again only for the purpose of logging? Isn't this exception going to be logged on upper level?